pandas/tseries/index.py

```diff
@@ -1202,9 +1202,10 @@ def intersection(self, other):
               not other.offset.isAnchored() or
               (not self.is_monotonic or not other.is_monotonic)):
             result = Index.intersection(self, other)
-            if isinstance(result, DatetimeIndex):
-                if result.freq is None:
-                    result.offset = to_offset(result.inferred_freq)
+            result = self._simple_new(result._values, name=result.name,
+                                      tz=result.tz)
+            if result.freq is None:
+                result.offset = to_offset(result.inferred_freq)
```

Comment:
Yes, as the simple result of Index.intersection inherits all attributes from other, and this is not the expected result. Therefore, it is necessary to create a new DatetimeIndex using only three of the Index.intersection attributes: _values, name and tz.